src/mongocxx/client.cpp

```diff
@@ -115,5 +118,17 @@ cursor client::list_databases() const {
     return cursor(result);
 }
 
+const client::impl& client::_get_impl() const {
+    if(!_impl) {
+        throw logic_error{make_error_code(error_code::k_invalid_client_object)};
+    }
+    return *_impl;
+}
+
+client::impl& client::_get_impl() {
+    auto cthis = const_cast<const client*>(this);
```

Comment:
you can use const with auto to capture, and then I don't think you'd need the const_case:
```
const auto const_this = this;
```